File: widgetssdk/src/main/java/com/glia/widgets/view/head/controller/ServiceChatHeadController.java
package com.glia.widgets.view.head.controller;

import android.view.View;

import androidx.core.util.Pair;

import com.glia.androidsdk.Operator;
import com.glia.androidsdk.comms.VisitorMediaState;
import com.glia.androidsdk.omnibrowse.OmnibrowseEngagement;
import com.glia.androidsdk.omnicore.OmnicoreEngagement;
import com.glia.widgets.UiTheme;
import com.glia.widgets.core.callvisualizer.domain.GliaOnCallVisualizerEndUseCase;
import com.glia.widgets.core.callvisualizer.domain.GliaOnCallVisualizerUseCase;
import com.glia.widgets.core.callvisualizer.domain.IsCallVisualizerScreenSharingUseCase;
import com.glia.widgets.core.chathead.domain.ResolveChatHeadNavigationUseCase;
import com.glia.widgets.core.chathead.domain.ToggleChatHeadServiceUseCase;
import com.glia.widgets.core.configuration.GliaSdkConfiguration;
import com.glia.widgets.core.engagement.domain.GetOperatorFlowableUseCase;
import com.glia.widgets.core.engagement.domain.GliaOnEngagementEndUseCase;
import com.glia.widgets.core.engagement.domain.GliaOnEngagementUseCase;
import com.glia.widgets.core.visitor.VisitorMediaUpdatesListener;
import com.glia.widgets.core.visitor.domain.AddVisitorMediaStateListenerUseCase;
import com.glia.widgets.core.visitor.domain.RemoveVisitorMediaStateListenerUseCase;
import com.glia.widgets.di.Dependencies;
import com.glia.widgets.helper.CommonExtensionsKt;
import com.glia.widgets.helper.Logger;
import com.glia.widgets.view.MessagesNotSeenHandler;
import com.glia.widgets.view.head.ChatHeadContract;
import com.glia.widgets.view.head.ChatHeadPosition;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

public class ServiceChatHeadController
    implements ChatHeadContract.Controller, VisitorMediaUpdatesListener,
    GliaOnEngagementEndUseCase.Listener, GliaOnCallVisualizerEndUseCase.Listener {
    private static final String TAG = ServiceChatHeadController.class.getSimpleName();

    private final ToggleChatHeadServiceUseCase toggleChatHeadServiceUseCase;
    private final ResolveChatHeadNavigationUseCase resolveChatHeadNavigationUseCase;

    private final GliaOnEngagementUseCase gliaOnEngagementUseCase;
    private final GliaOnCallVisualizerUseCase gliaOnCallVisualizerUseCase;
    private final GliaOnEngagementEndUseCase gliaOnEngagementEndUseCase;
    private final GliaOnCallVisualizerEndUseCase gliaOnCallVisualizerEndUseCase;
    private final MessagesNotSeenHandler messagesNotSeenHandler;
    private final AddVisitorMediaStateListenerUseCase addVisitorMediaStateListenerUseCase;
    private final RemoveVisitorMediaStateListenerUseCase removeVisitorMediaStateListenerUseCase;
    private final GetOperatorFlowableUseCase getOperatorFlowableUseCase;
    private final IsCallVisualizerScreenSharingUseCase isCallVisualizerScreenSharingUseCase;
    private final ChatHeadPosition chatHeadPosition;


    private final CompositeDisposable engagementDisposables = new CompositeDisposable();

    private ChatHeadContract.View chatHeadView;
    private State state = State.ENDED;
    private String operatorProfileImgUrl = null;
    private int unreadMessagesCount = 0;
    private boolean isOnHold = false;

    private GliaSdkConfiguration sdkConfiguration;
    private UiTheme buildTimeTheme;

    /*
     * We need to keep track of the currently active (topmost) view. This can be either ChatView
     * or CallView. CallView has translucent theme and this changes the usual lifecycle of an activity.
     * When the translucent CallActivity is on top of the ChatActivity and config change happens (e.g screen rotation)
     * then ChatActivity onResume and onPause are called right after CallActivity's onResume. Current
     * solution ignores such onResume calls because another activity is already in resumed state.
     */
    private String resumedViewName = null;
    private Disposable operatorDisposable = null;

    public ServiceChatHeadController(
        ToggleChatHeadServiceUseCase toggleChatHeadServiceUseCase,
        ResolveChatHeadNavigationUseCase resolveChatHeadNavigationUseCase,
        GliaOnEngagementUseCase gliaOnEngagementUseCase,
        GliaOnCallVisualizerUseCase gliaOnCallVisualizerUseCase,
        GliaOnEngagementEndUseCase onEngagementEndUseCase,
        GliaOnCallVisualizerEndUseCase onCallVisualizerEndUseCase,
        MessagesNotSeenHandler messagesNotSeenHandler,
        AddVisitorMediaStateListenerUseCase addVisitorMediaStateListenerUseCase,
        RemoveVisitorMediaStateListenerUseCase removeVisitorMediaStateListenerUseCase,
        ChatHeadPosition chatHeadPosition,
        GetOperatorFlowableUseCase getOperatorFlowableUseCase,
        IsCallVisualizerScreenSharingUseCase isCallVisualizerScreenSharingUseCase
    ) {
        this.toggleChatHeadServiceUseCase = toggleChatHeadServiceUseCase;
        this.resolveChatHeadNavigationUseCase = resolveChatHeadNavigationUseCase;
        this.gliaOnEngagementUseCase = gliaOnEngagementUseCase;
        this.gliaOnCallVisualizerUseCase = gliaOnCallVisualizerUseCase;
        this.gliaOnEngagementEndUseCase = onEngagementEndUseCase;
        this.gliaOnCallVisualizerEndUseCase = onCallVisualizerEndUseCase;
        this.messagesNotSeenHandler = messagesNotSeenHandler;
        this.chatHeadPosition = chatHeadPosition;
        this.addVisitorMediaStateListenerUseCase = addVisitorMediaStateListenerUseCase;
        this.removeVisitorMediaStateListenerUseCase = removeVisitorMediaStateListenerUseCase;
        this.getOperatorFlowableUseCase = getOperatorFlowableUseCase;
        this.isCallVisualizerScreenSharingUseCase = isCallVisualizerScreenSharingUseCase;
    }

    @Override
    public void onDestroy() {
        toggleChatHeadServiceUseCase.onDestroy();
        removeVisitorMediaStateListenerUseCase.execute(this);
        gliaOnEngagementEndUseCase.unregisterListener(this);
        gliaOnCallVisualizerEndUseCase.unregisterListener(this);
    }

    @Override
    public void onResume(View view) {
        setResumedViewName(view);

        // see the comment on resumedViewName field declaration above
        if (!isResumedView(view)) return;

        toggleChatHeadServiceUseCase.invoke(view.getClass().getSimpleName());
    }

    public void onPause(View view) {
        clearResumedViewName(view);
    }

    @Override
    public void onSetChatHeadView(ChatHeadContract.View view) {
        this.chatHeadView = view;
    }

    @Override
    public void onApplicationStop() {
        Logger.d(TAG, "onApplicationStop()");
        toggleChatHeadServiceUseCase.invoke(null);
    }

    @Override
    public void onChatHeadPositionChanged(int x, int y) {
        chatHeadPosition.set(x, y);
    }

    @Override
    public Pair<Integer, Integer> getChatHeadPosition() {
        return chatHeadPosition.get();
    }

    @Override
    public void onChatHeadClicked() {
        switch (resolveChatHeadNavigationUseCase.execute()) {
            case CALL_VIEW:
                chatHeadView.navigateToCall();
                break;
            case SCREEN_SHARING:
                chatHeadView.navigateToEndScreenSharing();
                break;
            case CHAT_VIEW:
            default:
                chatHeadView.navigateToChat();
        }
    }

    @Override
    public void onNewVisitorMediaState(VisitorMediaState visitorMediaState) {
    } // no-op

    @Override
    public void onHoldChanged(boolean isOnHold) {
        this.isOnHold = isOnHold;
        updateChatHeadView();
    }

    public void setSdkConfiguration(GliaSdkConfiguration configuration) {
        this.sdkConfiguration = configuration;
    }

    public void init() {
        gliaOnEngagementUseCase.execute(this::newEngagementLoaded);
        gliaOnCallVisualizerUseCase.invoke(this::newCallVisualizerEngagementLoaded);
        messagesNotSeenHandler.addListener(this::onUnreadMessageCountChange);
        addVisitorMediaStateListenerUseCase.execute(this);
    }

    public void updateChatHeadView() {
        if (chatHeadView != null && buildTimeTheme != null) {
            updateChatHeadViewState();
            updateOnHold();
            chatHeadView.showUnreadMessageCount(unreadMessagesCount);
            chatHeadView.updateConfiguration(buildTimeTheme, sdkConfiguration);
        }
    }

    public void setBuildTimeTheme(UiTheme theme) {
        this.buildTimeTheme = theme;
    }

    @Override
    public void engagementEnded() {
        callVisualizerEngagementEnded();
    }

    @Override
    public void callVisualizerEngagementEnded() {
        state = State.ENDED;
        operatorProfileImgUrl = null;
        unreadMessagesCount = 0;
        engagementDisposables.clear();
        updateChatHeadView();
    }

    private void newEngagementLoaded(OmnicoreEngagement engagement) {
        state = State.ENGAGEMENT;
        if (operatorDisposable != null) operatorDisposable.dispose();
        operatorDisposable = getOperatorFlowableUseCase.execute()
            .subscribe(
                this::operatorDataLoaded,
                throwable -> Logger.e(TAG, "getOperatorFlowableUseCase error: " + throwable.getMessage())
            );
        engagementDisposables.add(operatorDisposable);
        gliaOnEngagementEndUseCase.execute(this);
        toggleChatHeadServiceUseCase.invoke(resumedViewName);
        if (sdkConfiguration == null) setSdkConfiguration(Dependencies.getSdkConfigurationManager().createWidgetsConfiguration());
        updateChatHeadView();
    }

    private void newCallVisualizerEngagementLoaded(OmnibrowseEngagement engagement) {
        state = State.ENGAGEMENT;
        if (operatorDisposable != null) operatorDisposable.dispose();
        operatorDisposable = getOperatorFlowableUseCase.execute()
            .subscribe(
                this::operatorDataLoaded,
                throwable -> Logger.e(TAG, "getOperatorFlowableUseCase error: " + throwable.getMessage())
            );
        engagementDisposables.add(operatorDisposable);
        // To recieve callback to engagementEnded() after Call Visualizer engagement ends
        gliaOnCallVisualizerEndUseCase.execute(this);
        if (sdkConfiguration == null) setSdkConfiguration(Dependencies.getSdkConfigurationManager().createWidgetsConfiguration());
        updateChatHeadView();
    }

    private void onUnreadMessageCountChange(int count) {
        unreadMessagesCount = count;
        updateChatHeadView();
    }

    private void operatorDataLoaded(Operator operator) {
        operatorProfileImgUrl = CommonExtensionsKt.getImageUrl(operator);
        updateChatHeadView();
    }

    private void updateChatHeadViewState() {
        switch (state) {
            case ENGAGEMENT:
                decideOnBubbleDesign();
                break;
            case QUEUEING:
                chatHeadView.showQueueing();
                break;
            case ENDED:
            default:
                chatHeadView.showPlaceholder();
        }
    }

    private void decideOnBubbleDesign() {
        if (isCallVisualizerScreenSharingUseCase.invoke()) {
            chatHeadView.showScreenSharing();
        } else if (operatorProfileImgUrl != null) {
            chatHeadView.showOperatorImage(operatorProfileImgUrl);
        } else {
            chatHeadView.showPlaceholder();
        }
    }

    private void setResumedViewName(View view) {
        if (resumedViewName == null) {
            resumedViewName = view.getClass().getSimpleName();
        }
    }

    private void clearResumedViewName(View view) {
        // On quick configuration changes view can be null
        if (view != null && isResumedView(view)) {
            resumedViewName = null;
        }
    }

    private boolean isResumedView(View view) {
        return resumedViewName != null &&
            resumedViewName.equals(view.getClass().getSimpleName());
    }

    private void updateOnHold() {
        if (isOnHold && state == State.ENGAGEMENT) {
            chatHeadView.showOnHold();
        } else {
            chatHeadView.hideOnHold();
        }
    }

    private enum State {
        ENDED,
        QUEUEING,
        ENGAGEMENT
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/view/head/controller/ServiceChatHeadController.kt
package com.glia.widgets.view.head.controller

import android.annotation.SuppressLint
import android.view.View
import com.glia.androidsdk.Operator
import com.glia.widgets.callbacks.OnResult
import com.glia.widgets.di.Dependencies
import com.glia.widgets.engagement.EndAction
import com.glia.widgets.engagement.domain.CurrentOperatorUseCase
import com.glia.widgets.engagement.domain.EngagementStateUseCase
import com.glia.widgets.engagement.domain.VisitorMediaUseCase
import com.glia.widgets.helper.Logger
import com.glia.widgets.helper.TAG
import com.glia.widgets.helper.imageUrl
import com.glia.widgets.internal.chathead.domain.DisplayBubbleOutsideAppUseCase
import com.glia.widgets.internal.chathead.domain.ResolveChatHeadNavigationUseCase
import com.glia.widgets.internal.chathead.domain.ResolveChatHeadNavigationUseCase.Destinations
import com.glia.widgets.view.MessagesNotSeenHandler
import com.glia.widgets.view.head.ChatHeadContract
import com.glia.widgets.view.head.ChatHeadLogger
import com.glia.widgets.view.head.ChatHeadPosition
import com.glia.widgets.engagement.State as EngagementState

//This is in fact a singleton
@SuppressLint("CheckResult")
internal class ServiceChatHeadController(
    private val displayBubbleOutsideAppUseCase: DisplayBubbleOutsideAppUseCase,
    private val resolveChatHeadNavigationUseCase: ResolveChatHeadNavigationUseCase,
    messagesNotSeenHandler: MessagesNotSeenHandler,
    private var _chatHeadPosition: ChatHeadPosition,
    engagementStateUseCase: EngagementStateUseCase,
    currentOperatorUseCase: CurrentOperatorUseCase,
    visitorMediaUseCase: VisitorMediaUseCase
) : ChatHeadContract.Controller {
    private var chatHeadView: ChatHeadContract.View? = null
    private var state = State.ENDED
    private var operator: Operator? = null
    private var unreadMessagesCount = 0
    private var isOnHold = false
    private var engagementEndCallback: OnResult<Int> = OnResult { unreadMessageCount ->
        // Bubble should stay on the screen if there are unread messages to let the visitor know about that
        if (unreadMessageCount == 0) {
            updateStateOnEngagementEnded()
            Dependencies.secureConversations.unSubscribeFromUnreadMessageCount(engagementEndCallback)
        }
    }

    /*
     * We need to keep track of the currently active (topmost) view. This can be either ChatView
     * or CallView. CallView has a translucent theme and this changes the usual lifecycle of an activity.
     * When the translucent CallActivity is on top of the ChatActivity and config change happens (e.g., screen rotation)
     * then ChatActivity onResume and onPause are called right after CallActivity's onResume. The Current
     * solution ignores such onResume calls because another activity is already in resumed state.
     */
    private var resumedViewName: String? = null

    override val chatHeadPosition: ChatHeadPosition get() = _chatHeadPosition

    init {
        engagementStateUseCase().subscribe(::handleEngagementState)
        currentOperatorUseCase().subscribe(::operatorDataLoaded)
        messagesNotSeenHandler.addListener(::onUnreadMessageCountChange)
        visitorMediaUseCase.onHoldState.subscribe(::onHoldChanged)
    }

    override fun onResume(view: View?) {
        setResumedViewName(view)

        // see the comment on the resumedViewName field declaration above
        if (!isResumedView(view)) return
        displayBubbleOutsideAppUseCase(view?.javaClass?.simpleName)
    }

    override fun onPause(gliaOrRootView: View?) {
        clearResumedViewName(gliaOrRootView)
    }

    override fun onSetChatHeadView(view: ChatHeadContract.View) {
        chatHeadView = view
    }

    override fun onApplicationStop() {
        Logger.d(TAG, "onApplicationStop()")
        displayBubbleOutsideAppUseCase(null)
    }

    override fun onChatHeadPositionChanged(x: Int, y: Int) {
        _chatHeadPosition = ChatHeadPosition(x, y)
    }

    override fun onChatHeadClicked() {
        when (resolveChatHeadNavigationUseCase.execute()) {
            Destinations.CALL_VIEW -> chatHeadView?.navigateToCall()
            Destinations.CHAT_VIEW -> chatHeadView?.navigateToChat()
        }
        ChatHeadLogger.logChatHeadClicked()
    }

    private fun onHoldChanged(isOnHold: Boolean) {
        this.isOnHold = isOnHold
        updateChatHeadView()
    }


    private fun handleEngagementState(state: EngagementState) {
        when (state) {
            is EngagementState.EngagementStarted -> {
                newEngagementLoaded()
            }

            is EngagementState.Update -> toggleChatHead()

            is EngagementState.EngagementEnded -> engagementEnded(state.endAction)
            is EngagementState.QueueUnstaffed,
            is EngagementState.UnexpectedErrorHappened,
            is EngagementState.QueueingCanceled -> {
                engagementEnded(null)
            }

            is EngagementState.Queuing,
            is EngagementState.PreQueuing -> {
                queueingStarted()
            }

            else -> {
                //no op
            }
        }
    }

    private fun toggleChatHead() {
        state = State.ENGAGEMENT
        displayBubbleOutsideAppUseCase(resumedViewName)
        updateChatHeadView()
    }

    override fun updateChatHeadView() {
        if (chatHeadView != null) {
            updateChatHeadViewState()
            updateOnHold()
            chatHeadView!!.showUnreadMessageCount(unreadMessagesCount)
        }
    }

    private fun engagementEnded(action: EndAction?) {
        when (action) {
            EndAction.Retain -> {
                Dependencies.secureConversations.subscribeToUnreadMessageCount(engagementEndCallback)
            }

            else -> updateStateOnEngagementEnded()
        }
    }

    private fun updateStateOnEngagementEnded() {
        displayBubbleOutsideAppUseCase.onDestroy()
        isOnHold = false
        state = State.ENDED
        operator = null
        unreadMessagesCount = 0
        updateChatHeadView()
        ChatHeadLogger.reset()
    }

    private fun newEngagementLoaded() {
        isOnHold = false
        state = State.ENGAGEMENT
        displayBubbleOutsideAppUseCase(resumedViewName)
        updateChatHeadView()
    }

    private fun queueingStarted() {
        state = State.QUEUEING
        displayBubbleOutsideAppUseCase(resumedViewName)
        updateChatHeadView()
    }

    private fun onUnreadMessageCountChange(count: Int) {
        unreadMessagesCount = count
        updateChatHeadView()
    }

    private fun operatorDataLoaded(operator: Operator) {
        this.operator = operator
        updateChatHeadView()
    }

    private fun updateChatHeadViewState() {
        when (state) {
            State.ENGAGEMENT -> decideOnBubbleDesign()
            State.QUEUEING -> chatHeadView?.showQueueing()
            State.ENDED -> chatHeadView?.showPlaceholder()
        }
    }

    private fun decideOnBubbleDesign() {
        val view = chatHeadView ?: return
        operator?.imageUrl?.also(view::showOperatorImage) ?: view.showPlaceholder()
        // Here we draw operator image if we have it, otherwise we show placeholder
        // This is the only place where we show operator image inside bubble, so this is an indicator that the operator is connected
        // This function is called multiple times during engagement, but we're filtering out duplicate logs inside ChatHeadLogger
        ChatHeadLogger.logOperatorConnected(operator ?: return)
    }

    private fun setResumedViewName(view: View?) {
        resumedViewName = view?.javaClass?.simpleName
    }

    private fun clearResumedViewName(view: View?) {
        // On quick configuration changes view can be null
        if (view != null && isResumedView(view)) {
            resumedViewName = null
        }
    }

    private fun isResumedView(view: View?): Boolean {
        return resumedViewName != null && resumedViewName == view?.javaClass?.simpleName
    }

    private fun updateOnHold() {
        if (isOnHold && state == State.ENGAGEMENT) {
            chatHeadView?.showOnHold()
        } else {
            chatHeadView?.hideOnHold()
        }
    }

    override fun onDestroy() {
        //no op
    }

    private enum class State {
        ENDED,
        QUEUEING,
        ENGAGEMENT
    }

}
